exporter/tanzuobservabilityexporter/metrics.go

```diff
@@ -345,3 +350,195 @@ func (s *sumConsumer) pushNumberDataPoint(
+	leInUse                  counter
+	noAggregationTemporality counter
+}
+
+// newHistogramReporting returns a new histogramReporting instance.
+// log messages are sent to logger.
+func newHistogramReporting(logger *zap.Logger) *histogramReporting {
+	return &histogramReporting{logger: logger}
+}
+
+// Malformed returns the number of malformed histogram data points.
+func (r *histogramReporting) Malformed() int64 {
+	return r.malformedHistograms.Get()
+}
+
+// LeTagInUse returns the number of histogram data points already using the 'le'
+// tag.
+func (r *histogramReporting) LeTagInUse() int64 {
+	return r.leInUse.Get()
+}
+
+// NoAggregationTemporality returns the number of histogram metrics that have no
+// aggregation temporality.
+func (r *histogramReporting) NoAggregationTemporality() int64 {
+	return r.noAggregationTemporality.Get()
+}
+
+// LogMalformed logs seeing one malformed data point.
+func (r *histogramReporting) LogMalformed(metric pdata.Metric) {
+	namef := zap.String(metricNameString, metric.Name())
+	r.logger.Debug("Malformed histogram", namef)
+	r.malformedHistograms.Inc()
+}
+
+// LogLeTagInUse logs seeing one data point using the 'le' tag
+func (r *histogramReporting) LogLeTagInUse(metric pdata.Metric) {
+	namef := zap.String(metricNameString, metric.Name())
+	r.logger.Debug("le tag already in use", namef)
+	r.leInUse.Inc()
+}
+
+// LogNoAggregationTemporality logs seeing a histogram metric with no aggregation temporality
+func (r *histogramReporting) LogNoAggregationTemporality(metric pdata.Metric) {
+	namef := zap.String(metricNameString, metric.Name())
+	r.logger.Debug("histogram metric missing aggregation temporality", namef)
+	r.noAggregationTemporality.Inc()
+}
+
+// Report sends the counts in this instancce to wavefront.
+// sender is what sends to wavefront. Any errors sending get added to errs.
+func (r *histogramReporting) Report(sender gaugeSender, errs *[]error) {
+	r.malformedHistograms.Report(malformedHistogramMetricName, nil, sender, errs)
+	r.leInUse.Report(leInUseMetricName, nil, sender, errs)
+	r.noAggregationTemporality.Report(
+		noAggregationTemporalityMetricName,
+		typeIsHistogramTags,
+		sender,
+		errs)
+}
+
+type histogramConsumer struct {
+	cumulative            histogramDataPointConsumer
+	delta                 histogramDataPointConsumer
+	sender                gaugeSender
+	reporting             *histogramReporting
+	reportInternalMetrics bool
+}
+
+// newHistogramConsumer returns a metricConsumer that consumes histograms.
+// cumulative and delta handle cumulative and delta histograms respectively.
+// sender sends internal metrics to wavefront. Caller can pass nil for
+// options to get the defaults.
+func newHistogramConsumer(
+	cumulative, delta histogramDataPointConsumer,
+	sender gaugeSender,
+	options *consumerOptions,
+) typedMetricConsumer {
+	var fixedOptions consumerOptions
+	if options != nil {
+		fixedOptions = *options
+	}
+	fixedOptions.replaceZeroFieldsWithDefaults()
```

Comment:
You can use this instead:

https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/51d2a7775b22c07beeb508612564ff9a483cd480/extension/jaegerremotesampling/factory.go#L39-L43

Then you tie it with the factory:
https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/51d2a7775b22c07beeb508612564ff9a483cd480/extension/jaegerremotesampling/factory.go#L35

When your component is created, this is used as the base for the config, with the options your users provided overriding the default values.

Reply:
What if I got rid of the consumerOptions type entirely?  It only has two fields: `Logger` and `ReportInternalMetrics`.  Any function that expects a `consumerOptions` will take only a zap.Logger. A nil logger means use the nop logger and don't report internal metrics; A non-nil logger means use that logger and report internal metrics.  Note that I would pass nil for logger only in test code. I think this is fine because I wasn't planning on letting the user decide whether or not to report internal metrics in the config file. I was planning on always reporting internal metrics. What do you think?

Reply:
Done. Got rid of consumerOptions type completely.